--- bison-2.3-orig/lib/stdbool_.h	2006-01-26 13:31:49.000000000 +0000
+++ bison-2.3/lib/stdbool_.h	2008-03-28 05:24:05.000000000 +0000
@@ -55,7 +55,7 @@
 
 /* BeOS <sys/socket.h> already #defines false 0, true 1.  We use the same
    definitions below, but temporarily we have to #undef them.  */
-#ifdef __BEOS__
+#if defined(__BEOS__) && !defined(__HAIKU__)
 # include <OS.h> /* defines bool but not _Bool */
 # undef false
 # undef true
@@ -70,7 +70,7 @@
    (see ISO C 99 6.7.2.2.(4)); however, '_Bool' must promote to 'int'
    (see ISO C 99 6.3.1.1.(2)).  So we add a negative value to the
    enum; this ensures that '_Bool' promotes to 'int'.  */
-#if defined __cplusplus || defined __BEOS__
+#if defined __cplusplus || (defined __BEOS__ && !defined(__HAIKU__))
   /* A compiler known to have 'bool'.  */
   /* If the compiler already has both 'bool' and '_Bool', we can assume they
      are the same types.  */
