diff -ur openssh-4.7p1-orig/includes.h openssh-4.7p1/includes.h
--- openssh-4.7p1-orig/includes.h	2007-06-11 06:44:03.000000000 +0200
+++ openssh-4.7p1/includes.h	2008-03-29 03:00:42.000000000 +0100
@@ -109,7 +109,9 @@
 #endif
 
 #include <netinet/in.h>
+#ifndef __HAIKU__
 #include <netinet/in_systm.h> /* For typedefs */
+#endif
 #ifdef HAVE_RPC_TYPES_H
 # include <rpc/types.h> /* For INADDR_LOOPBACK */
 #endif
